### PR TITLE
Use ElasticSearch RestClient instead of vanilla httpclient (Merge into develop)

### PR DESCRIPTION
**Pull request by bsura**
Advantages of using ES Rest Client:
- minimal dependencies
- load balancing across all available nodes
- failover in case of node failures and upon specific response codes
- failed connection penalization (whether a failed node is retried depends on how many consecutive times it failed; the more failed attempts the longer the client will wait before trying that same node again)
- persistent connections
- trace logging of requests and responses
- optional automatic discovery of cluster nodes

@rajsarkapally-sfdc @dilipdevaraj-sfdc

### CODE REVIEW

**Comment by bsura** [comment]
@dilipdevaraj-sfdc Please review.

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java [hunk 1]
```diff
@@ -65,7 +65,8 @@
 	private static final String INDEX_NAME = "metadata_index";
 	private static final String TYPE_NAME = "metadata_type";
 	private static final String KEEP_SCROLL_CONTEXT_OPEN_FOR = "1m";
-	private static final int INDEX_MAX_RESULT_WINDOW = 10000; 
+	private static final int INDEX_MAX_RESULT_WINDOW = 10000;
+	private static final int NUMBER_OF_REPLICAS = 3;
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java]: What is the advantage in increasing replicas to 3. When it is 2, we will still have original index, and then 2 more copies ?

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/DefaultTSDBService.java [hunk 2]
```diff
@@ -447,23 +443,21 @@ private ObjectMapper getMapper() {
...
+    private CloseableHttpClient getClient(int connCount, int connTimeout, int socketTimeout, String...endpoints) throws MalformedURLException {
         PoolingHttpClientConnectionManager connMgr = new PoolingHttpClientConnectionManager();
-
         connMgr.setMaxTotal(connCount);
-        connMgr.setDefaultMaxPerRoute(connCount);
-
-        String route = endpoint.substring(0, endpoint.lastIndexOf(":"));
-        HttpHost host = new HttpHost(route, port);
+        
+        for(String endpoint : endpoints) {
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/DefaultTSDBService.java]: What is the basis for the change ?
Previously we had,  1 write http client  for  every tsdb write endpoint.
Now we are cycling through all write endpoints but adding this configuration to a single write http client.

Is this so that we can support a large number of tsdb endpoints (say >100)  per commit client, so round robin of commits works without any throughput reduction ?

File: ArgusSDK/src/main/java/com/salesforce/dva/argus/sdk/entity/MetricSchemaRecord.java [hunk 3]
```diff
@@ -143,16 +143,16 @@ public void setMetric(String metric) {
...
      */
     public String getTagKey() {
-        return tagKey;
+        return tagk;
     }
 
     /**
      * Specifies the tag key associated with the result.
      *
      * @param  tagKey  The tag key. Can be null.
```
Comment (dilipdevaraj-sfdc) [inline ArgusSDK/src/main/java/com/salesforce/dva/argus/sdk/entity/MetricSchemaRecord.java]: change name in Javadoc

File: ArgusSDK/src/main/java/com/salesforce/dva/argus/sdk/entity/MetricSchemaRecord.java [hunk 4]
```diff
@@ -161,16 +161,16 @@ public void setTagKey(String tagKey) {
...
      */
     public String getTagValue() {
-        return tagValue;
+        return tagv;
     }
 
     /**
      * Specifies the tag value associated with the result.
      *
      * @param  tagValue  The tag value. Can be null.
```
Comment (dilipdevaraj-sfdc) [inline ArgusSDK/src/main/java/com/salesforce/dva/argus/sdk/entity/MetricSchemaRecord.java]: change name in Javadoc

File: ArgusSDK/src/main/java/com/salesforce/dva/argus/sdk/entity/MetricSchemaRecord.java (hunk 3 above)
Comment (bsura) [inline ArgusSDK/src/main/java/com/salesforce/dva/argus/sdk/entity/MetricSchemaRecord.java]: Done.

File: ArgusSDK/src/main/java/com/salesforce/dva/argus/sdk/entity/MetricSchemaRecord.java (hunk 4 above)
Comment (bsura) [inline ArgusSDK/src/main/java/com/salesforce/dva/argus/sdk/entity/MetricSchemaRecord.java]: Done.

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java (hunk 1 above)
Comment (bsura) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java]: Done.

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/DefaultTSDBService.java (hunk 2 above)
Comment (bsura) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/DefaultTSDBService.java]: Let's discuss this in person.

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/DefaultTSDBService.java [hunk 5]
```diff
@@ -447,23 +443,21 @@ private ObjectMapper getMapper() {
...
+        	URL url = new URL(endpoint);
+            int port = url.getPort();
+            requireArgument(port != -1, "TSDB endpoint must include explicit port.");
+            HttpHost host = new HttpHost(url.getHost(),	url.getPort());
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/DefaultTSDBService.java]: Can remove extra white spaces and indent code in files

File: ArgusWeb/app/js/directives/charts/chart.js [hunk 6]
```diff
@@ -7,7 +7,8 @@ angular.module('argus.directives.charts.chart', [])
 		var chartNameIndex = 1;
 		function compileLineChart(scope, newChartId, series, dateConfig, updatedOptionList) {
 			// empty any previous content
-			$('#' + newChartId).empty();
+			angular.element('#' + newChartId).empty();
```
Comment (dilipdevaraj-sfdc) [inline ArgusWeb/app/js/directives/charts/chart.js]: Why are unrelated UI changes showing up in this PR?

File: ArgusWebServices/src/main/java/com/salesforce/dva/argus/ws/dto/MetricDiscoveryQueryDto.java [hunk 7]
```diff
@@ -112,33 +110,33 @@ public void setKeywordQueryNative(boolean isKeywordQueryNative) {
...
+			anotherQuery.setNamespace("namespace");
+			anotherQuery.setScope("scope");
+			anotherQuery.setMetric("metric");
+			anotherQuery.setTagk("tagKey");
+			anotherQuery.setTagv("tagValue");
+			anotherQuery.setLimit(10);
+			anotherQuery.setType(RecordType.SCOPE);
+			anotherQuery.setPage(1);
 			
 			MetricDiscoveryQueryDto yetAnotherResult = new MetricDiscoveryQueryDto();
```
Comment (dilipdevaraj-sfdc) [inline ArgusWebServices/src/main/java/com/salesforce/dva/argus/ws/dto/MetricDiscoveryQueryDto.java]: change here too the  object name.

**Comment by dilipdevaraj-sfdc** [comment]
Are there any UI related changes that need to be done. If so will that be part of a separate PR ?

**Comment by dilipdevaraj-sfdc** [comment]
Wrt overall comment of  "Advantages of using ES Rest Client:"
- Will the shards of a index be evenly distributed to as many nodes as possible ?
- When the querying happens does the client know which specific nodes will have the data or does it have to query all nodes ?

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java [hunk 8]
```diff
@@ -388,14 +521,19 @@ private String _constructTermQuery(MetricSchemaRecordQuery query, int from, int
 		return rootNode.toString();
 	}
 
-	
+	/*
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java]: Can we get rid of the unused method ?

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java [hunk 9]
```diff
@@ -239,105 +285,189 @@ protected void implementationSpecificPut(List<Metric> metrics) {
...
 			}
-		} catch (UnsupportedEncodingException e) {
-			throw new SystemException("Search failed.", e);
+		} catch (IOException e) {
+			throw new SystemException(e);
 		}
 	}
 	
 	@Override
 	public List<MetricSchemaRecord> keywordSearch(KeywordQuery kq) {
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java]: Why is this method named keywordSearch ?
In ES search of String fields we either use FULL_TEXT(Tokenized) or KEYWORD(No tokenization) based search.

In the if clause you are searching using raw query, and in else clause you are using the individual fields of scope, metric...

Can we simplify this method/ break it into smaller methods ?

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java [hunk 10]
```diff
@@ -145,35 +181,38 @@ protected void implementationSpecificPut(List<Metric> metrics) {
...
         boolean scroll = false;
-		StringBuilder sb = new StringBuilder(_esEndpoint).append("/")
-														  .append(INDEX_NAME)
-														  .append("/")
-														  .append(TYPE_NAME)
-														  .append("/")
-														  .append("_search");
-		
-		int from = 0, size;
+		StringBuilder sb = new StringBuilder().append("/")
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java]: Can you indent this code

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java (hunk 9 above)
Comment (bsura) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java]: I will change the method name to make it more expressive. Maybe tokenizedSearch or fullTextSearch?

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java (hunk 8 above)
Comment (bsura) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/schema/ElasticSearchSchemaService.java]: Will do.

**Comment by xizi-xu** [comment]
@bsura @dilipdevaraj-sfdc you guys want to merge this?

**Comment by bsura** [comment]
@xizi-xu Done.
